In [545]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from geopy.geocoders import Nominatim
from geopy.exc import GeocoderTimedOut
from IPython.core.display import display, HTML
import time

In [546]:
df1 = pd.read_csv('zomato_data.csv')

In [547]:
df1.head()

online_order book_table   rate  votes            rest_type  \
0          Yes        Yes  4.1/5    775        Casual Dining   
1          Yes         No  4.1/5    787        Casual Dining   
2          Yes         No  3.8/5    918  Cafe, Casual Dining   
3           No         No  3.7/5     88          Quick Bites   
4           No         No  3.8/5    166        Casual Dining   

                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines approx_costfor_two_people listed_intype  \
0  North Indian, Mughlai, Chinese                       800        Buffet   
1     Chinese, North Indian, Thai                       800        Buffet   
2          Cafe, Mexican, Italian                       800        Buffet   
3      South Indian, North Indian                       300        Buffet   
4        North Indian, Rajasthani                       600        Buffet   

  listed_incity  
0  Banashankari  
1  Banashankari  
2  Banashankari  
3  Banashankari  
4  Banashankari

In [548]:
df1.shape

(51717, 10)

Inference : The Dataset zomato_data.csv contains 51717 rows and 10 columns.

In [549]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   online_order               51717 non-null  object
 1   book_table                 51717 non-null  object
 2   rate                       43942 non-null  object
 3   votes                      51717 non-null  int64 
 4   rest_type                  51490 non-null  object
 5   dish_liked                 23639 non-null  object
 6   cuisines                   51672 non-null  object
 7   approx_costfor_two_people  51371 non-null  object
 8   listed_intype              51717 non-null  object
 9   listed_incity              51717 non-null  object
dtypes: int64(1), object(9)
memory usage: 3.9+ MB


In [550]:
df2 = pd.read_csv('Geographical Coordinates.csv')

In [551]:
df2.head()

listed_incity   Latitude  Longitude
0       Banashankari  12.939333  77.553982
1  Bannerghatta Road  12.952660  77.605048
2       Basavanagudi  12.941726  77.575502
3          Bellandur  12.925352  77.675941
4       Brigade Road  12.967358  77.606435

In [552]:
df2.shape

(26, 3)

Inference : The Dataset Geographical Coordinates.csv contains 51717 rows and 10 columns.

In [553]:
# Checking Info of the data
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26 entries, 0 to 25
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   listed_incity  26 non-null     object 
 1   Latitude       26 non-null     float64
 2   Longitude      26 non-null     float64
dtypes: float64(2), object(1)
memory usage: 756.0+ bytes


Data Cleaning and Preprocessing

In [554]:
# Checking Info of the data
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   online_order               51717 non-null  object
 1   book_table                 51717 non-null  object
 2   rate                       43942 non-null  object
 3   votes                      51717 non-null  int64 
 4   rest_type                  51490 non-null  object
 5   dish_liked                 23639 non-null  object
 6   cuisines                   51672 non-null  object
 7   approx_costfor_two_people  51371 non-null  object
 8   listed_intype              51717 non-null  object
 9   listed_incity              51717 non-null  object
dtypes: int64(1), object(9)
memory usage: 3.9+ MB


In [555]:
# Checking number of null values present in columns
df1.isnull().sum()

online_order                     0
book_table                       0
rate                          7775
votes                            0
rest_type                      227
dish_liked                   28078
cuisines                        45
approx_costfor_two_people      346
listed_intype                    0
listed_incity                    0
dtype: int64

Inference : Column rate, rest_type, dish_liked, cuisines, approx_costfor_two_people contain null values as 7775, 227, 28078, 45, 346 respectively.

In [556]:
# Describing the data
df1.describe().T

count        mean         std  min  25%   50%    75%      max
votes  51717.0  283.697527  803.838853  0.0  7.0  41.0  198.0  16832.0

Step 1 : Rating Column (rate)

In [557]:
# Checking Number of Null values in 'rate' column
print('Null values in \'rate\' column is' ,df1['rate'].isnull().sum())

Null values in 'rate' column is 7775


In [558]:
# Replacing Null Values by '-' symbol
df1['rate'] = df1['rate'].fillna('-')

In [559]:
# Checking head/tail for confirmation of filling '-' value
df1['rate'].tail()

51712    3.6 /5
51713         -
51714         -
51715    4.3 /5
51716    3.4 /5
Name: rate, dtype: object

In [560]:
# Replaceing '/5' by '' so that rating remains numerical value
df1['rate'] = df1['rate'].str.replace('/5', '')

In [561]:
# Checking head/tail
df1['rate'].tail()

51712    3.6 
51713       -
51714       -
51715    4.3 
51716    3.4 
Name: rate, dtype: object

In [562]:
# Converting it into numeric value
df1['rate'] = pd.to_numeric(df1['rate'], errors = 'coerce') # Here errors = 'coerce' is used so that '-' string is parsed easily

In [563]:
# Checking head/tail
df1['rate'].tail()

51712    3.6
51713    NaN
51714    NaN
51715    4.3
51716    3.4
Name: rate, dtype: float64

In [564]:
# Checking type of 'rate' column
print('Type of \'rate\' column is :', df1['rate'].dtype)

Type of 'rate' column is : float64


In [565]:
# Finding Median of 'rate' column
Med = df1['rate'].median()
print('Median of \'rate\' column is :', Med)

Median of 'rate' column is : 3.7


In [566]:
# Filling Nul Values with median value 3.7
df1['rate'] = df1['rate'].fillna(df1['rate'].median())

In [567]:
# Checking head/tail
df1['rate'].tail()

51712    3.6
51713    3.7
51714    3.7
51715    4.3
51716    3.4
Name: rate, dtype: float64

In [568]:
# Checking Number of Null values in 'rate' column
print('Null values in \'rate\' column is' ,df1['rate'].isnull().sum())

Null values in 'rate' column is 0


In [569]:
# Checking Info of the data
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   online_order               51717 non-null  object 
 1   book_table                 51717 non-null  object 
 2   rate                       51717 non-null  float64
 3   votes                      51717 non-null  int64  
 4   rest_type                  51490 non-null  object 
 5   dish_liked                 23639 non-null  object 
 6   cuisines                   51672 non-null  object 
 7   approx_costfor_two_people  51371 non-null  object 
 8   listed_intype              51717 non-null  object 
 9   listed_incity              51717 non-null  object 
dtypes: float64(1), int64(1), object(8)
memory usage: 3.9+ MB


In [570]:
# Checking number of null values present in columns
df1.isnull().sum()

online_order                     0
book_table                       0
rate                             0
votes                            0
rest_type                      227
dish_liked                   28078
cuisines                        45
approx_costfor_two_people      346
listed_intype                    0
listed_incity                    0
dtype: int64

In [571]:
# Describing the data
df1.describe().T

count        mean         std  min  25%   50%    75%      max
rate   51717.0    3.700362    0.395391  1.8  3.5   3.7    3.9      4.9
votes  51717.0  283.697527  803.838853  0.0  7.0  41.0  198.0  16832.0

Step 2 : Cost Column (approx_costfor_two_people)

In [572]:
# Checking head/tail of column
df1['approx_costfor_two_people'].tail()

51712    1,500
51713      600
51714    2,000
51715    2,500
51716    1,500
Name: approx_costfor_two_people, dtype: object

In [573]:
# Removing commas from column values
df1['approx_costfor_two_people'] = df1['approx_costfor_two_people'].str.replace(',', '')

In [574]:
# Checking head/tail of column
df1['approx_costfor_two_people'].tail()

51712    1500
51713     600
51714    2000
51715    2500
51716    1500
Name: approx_costfor_two_people, dtype: object

In [575]:
# Checking Number of Null values in 'approx_costfor_two_people' column
print('Null values in \'approx_costfor_two_people\' column is' ,df1['approx_costfor_two_people'].isnull().sum())

Null values in 'approx_costfor_two_people' column is 346


In [576]:
# Converting it into numeric value
df1['approx_costfor_two_people'] = pd.to_numeric(df1['approx_costfor_two_people'])

In [577]:
# Checking head/tail of column
df1['approx_costfor_two_people'].tail()

51712    1500.0
51713     600.0
51714    2000.0
51715    2500.0
51716    1500.0
Name: approx_costfor_two_people, dtype: float64

In [578]:
# Checking type of 'rate' column
print('Type of \'rate\' column is :', df1['approx_costfor_two_people'].dtype)

Type of 'rate' column is : float64


In [579]:
# Finding Median of 'approx_costfor_two_people' column
Med = df1['approx_costfor_two_people'].median()
print('Median of \'approx_costfor_two_people\' column is :', Med)

Median of 'approx_costfor_two_people' column is : 400.0


In [580]:
# Filling Nul Values with median value 400.0
df1['approx_costfor_two_people'] = df1['approx_costfor_two_people'].fillna(df1['approx_costfor_two_people'].median())

In [581]:
# Checking head/tail of column
df1['approx_costfor_two_people'].tail()

51712    1500.0
51713     600.0
51714    2000.0
51715    2500.0
51716    1500.0
Name: approx_costfor_two_people, dtype: float64

In [582]:
# Checking Number of Null values in 'approx_costfor_two_people' column
print('Null values in \'approx_costfor_two_people\' column is' ,df1['approx_costfor_two_people'].isnull().sum())

Null values in 'approx_costfor_two_people' column is 0


In [583]:
# Checking Info of the data
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   online_order               51717 non-null  object 
 1   book_table                 51717 non-null  object 
 2   rate                       51717 non-null  float64
 3   votes                      51717 non-null  int64  
 4   rest_type                  51490 non-null  object 
 5   dish_liked                 23639 non-null  object 
 6   cuisines                   51672 non-null  object 
 7   approx_costfor_two_people  51717 non-null  float64
 8   listed_intype              51717 non-null  object 
 9   listed_incity              51717 non-null  object 
dtypes: float64(2), int64(1), object(7)
memory usage: 3.9+ MB


In [584]:
# Checking number of null values present in columns
df1.isnull().sum()

online_order                     0
book_table                       0
rate                             0
votes                            0
rest_type                      227
dish_liked                   28078
cuisines                        45
approx_costfor_two_people        0
listed_intype                    0
listed_incity                    0
dtype: int64

In [585]:
# Describing the data
df1.describe().T

count        mean         std   min    25%  \
rate                       51717.0    3.700362    0.395391   1.8    3.5   
votes                      51717.0  283.697527  803.838853   0.0    7.0   
approx_costfor_two_people  51717.0  554.391689  437.563723  40.0  300.0   

                             50%    75%      max  
rate                         3.7    3.9      4.9  
votes                       41.0  198.0  16832.0  
approx_costfor_two_people  400.0  650.0   6000.0

Step 3: Categorical Columns

Step 3.1: dish_liked → Replace NaN with "Not Available"

In [586]:
# Checking how many Null values are there in 'dish_liked' column
print('Null values in \'dish_liked\' column is', df1['dish_liked'].isnull().sum())

Null values in 'dish_liked' column is 28078


In [587]:
# Checking head/tail of column
df1['dish_liked'].tail()

51712                             NaN
51713                             NaN
51714                             NaN
51715    Cocktails, Pizza, Buttermilk
51716                             NaN
Name: dish_liked, dtype: object

In [588]:
# Replacing Null values by 'Not Available'
df1['dish_liked'] = df1['dish_liked'].fillna('Not Available')

In [589]:
# Checking head/tail of column
df1['dish_liked'].tail()

51712                   Not Available
51713                   Not Available
51714                   Not Available
51715    Cocktails, Pizza, Buttermilk
51716                   Not Available
Name: dish_liked, dtype: object

In [590]:
# Checking how many Null values are there in 'dish_liked' column
print('Null values in \'dish_liked\' column is', df1['dish_liked'].isnull().sum())

Null values in 'dish_liked' column is 0


In [591]:
# Checking Info of the data
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   online_order               51717 non-null  object 
 1   book_table                 51717 non-null  object 
 2   rate                       51717 non-null  float64
 3   votes                      51717 non-null  int64  
 4   rest_type                  51490 non-null  object 
 5   dish_liked                 51717 non-null  object 
 6   cuisines                   51672 non-null  object 
 7   approx_costfor_two_people  51717 non-null  float64
 8   listed_intype              51717 non-null  object 
 9   listed_incity              51717 non-null  object 
dtypes: float64(2), int64(1), object(7)
memory usage: 3.9+ MB


In [592]:
# Checking number of null values present in columns
df1.isnull().sum()

online_order                   0
book_table                     0
rate                           0
votes                          0
rest_type                    227
dish_liked                     0
cuisines                      45
approx_costfor_two_people      0
listed_intype                  0
listed_incity                  0
dtype: int64

In [593]:
# Describing the data
df1.describe().T

count        mean         std   min    25%  \
rate                       51717.0    3.700362    0.395391   1.8    3.5   
votes                      51717.0  283.697527  803.838853   0.0    7.0   
approx_costfor_two_people  51717.0  554.391689  437.563723  40.0  300.0   

                             50%    75%      max  
rate                         3.7    3.9      4.9  
votes                       41.0  198.0  16832.0  
approx_costfor_two_people  400.0  650.0   6000.0

Step 3.2: cuisines → Replace NaN with "Other"

In [594]:
# Checking how many Null values are there in 'cuisines' column
print('Null values in \'cuisines\' column is', df1['cuisines'].isnull().sum())

Null values in 'cuisines' column is 45


In [595]:
# Checking head/tail of column
df1['cuisines'].tail()

51712                               Continental
51713                               Finger Food
51714                               Finger Food
51715                               Finger Food
51716    Finger Food, North Indian, Continental
Name: cuisines, dtype: object

In [596]:
# Replacing Null values by 'Other'
df1['cuisines'] = df1['cuisines'].fillna('')

In [597]:
# Checking head/tail of column
df1['cuisines'].tail()

51712                               Continental
51713                               Finger Food
51714                               Finger Food
51715                               Finger Food
51716    Finger Food, North Indian, Continental
Name: cuisines, dtype: object

In [598]:
# Checking how many Null values are there in 'cuisines' column
print('Null values in \'cuisines\' column is', df1['cuisines'].isnull().sum())

Null values in 'cuisines' column is 0


In [599]:
# Checking Info of the data
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   online_order               51717 non-null  object 
 1   book_table                 51717 non-null  object 
 2   rate                       51717 non-null  float64
 3   votes                      51717 non-null  int64  
 4   rest_type                  51490 non-null  object 
 5   dish_liked                 51717 non-null  object 
 6   cuisines                   51717 non-null  object 
 7   approx_costfor_two_people  51717 non-null  float64
 8   listed_intype              51717 non-null  object 
 9   listed_incity              51717 non-null  object 
dtypes: float64(2), int64(1), object(7)
memory usage: 3.9+ MB


In [600]:
# Checking number of null values present in columns
df1.isnull().sum()

online_order                   0
book_table                     0
rate                           0
votes                          0
rest_type                    227
dish_liked                     0
cuisines                       0
approx_costfor_two_people      0
listed_intype                  0
listed_incity                  0
dtype: int64

In [601]:
# Describing the data
df1.describe().T

count        mean         std   min    25%  \
rate                       51717.0    3.700362    0.395391   1.8    3.5   
votes                      51717.0  283.697527  803.838853   0.0    7.0   
approx_costfor_two_people  51717.0  554.391689  437.563723  40.0  300.0   

                             50%    75%      max  
rate                         3.7    3.9      4.9  
votes                       41.0  198.0  16832.0  
approx_costfor_two_people  400.0  650.0   6000.0

Step 3.3: rest_type → Replace NaN with "Unknown"

In [602]:
# Checking how many Null values are there in 'rest_type' column
print('Null values in \'rest_type\' clumn is', df1['rest_type'].isnull().sum())

Null values in 'rest_type' clumn is 227


In [603]:
# Checking head/tail of column
df1['rest_type'].tail()

51712                   Bar
51713                   Bar
51714                   Bar
51715                   Bar
51716    Bar, Casual Dining
Name: rest_type, dtype: object

In [604]:
# Replacing Null values by 'Unknown'
df1['rest_type'] = df1['rest_type'].fillna('Unknown')

In [605]:
# Checking head/tail of column
df1['rest_type'].tail()

51712                   Bar
51713                   Bar
51714                   Bar
51715                   Bar
51716    Bar, Casual Dining
Name: rest_type, dtype: object

In [606]:
# Checking how many Null values are there in 'rest_type' column
print('Null values in \'rest_type\' clumn is', df1['rest_type'].isnull().sum())

Null values in 'rest_type' clumn is 0


In [607]:
# Checking Info of the data
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   online_order               51717 non-null  object 
 1   book_table                 51717 non-null  object 
 2   rate                       51717 non-null  float64
 3   votes                      51717 non-null  int64  
 4   rest_type                  51717 non-null  object 
 5   dish_liked                 51717 non-null  object 
 6   cuisines                   51717 non-null  object 
 7   approx_costfor_two_people  51717 non-null  float64
 8   listed_intype              51717 non-null  object 
 9   listed_incity              51717 non-null  object 
dtypes: float64(2), int64(1), object(7)
memory usage: 3.9+ MB


In [608]:
# Checking number of null values present in columns
df1.isnull().sum()

online_order                 0
book_table                   0
rate                         0
votes                        0
rest_type                    0
dish_liked                   0
cuisines                     0
approx_costfor_two_people    0
listed_intype                0
listed_incity                0
dtype: int64

In [609]:
# Describing the data
df1.describe().T

count        mean         std   min    25%  \
rate                       51717.0    3.700362    0.395391   1.8    3.5   
votes                      51717.0  283.697527  803.838853   0.0    7.0   
approx_costfor_two_people  51717.0  554.391689  437.563723  40.0  300.0   

                             50%    75%      max  
rate                         3.7    3.9      4.9  
votes                       41.0  198.0  16832.0  
approx_costfor_two_people  400.0  650.0   6000.0

Step 4: Votes Column

In [610]:
# Checking how many Null values are there in 'votes' column
print('Null values in \'votes\' clumn is', df1['votes'].isnull().sum())

Null values in 'votes' clumn is 0


Inference : Ideally, there is no need of filling null values in 'votes' column since it has 0 null values. But for formality given in instructions, I am doing this.

In [611]:
# Finding Median of 'votes' column
Med = df1['votes'].median()
print('Median of \'votes\' column is :', Med)

Median of 'votes' column is : 41.0


In [612]:
# Filling Nul Values with median value 3.7
df1['votes'] = df1['votes'].fillna(df1['votes'].median())

In [613]:
# Checking how many Null values are there in 'votes' column
print('Null values in \'votes\' clumn is', df1['votes'].isnull().sum())

Null values in 'votes' clumn is 0


In [614]:
# Checking Info of the data
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   online_order               51717 non-null  object 
 1   book_table                 51717 non-null  object 
 2   rate                       51717 non-null  float64
 3   votes                      51717 non-null  int64  
 4   rest_type                  51717 non-null  object 
 5   dish_liked                 51717 non-null  object 
 6   cuisines                   51717 non-null  object 
 7   approx_costfor_two_people  51717 non-null  float64
 8   listed_intype              51717 non-null  object 
 9   listed_incity              51717 non-null  object 
dtypes: float64(2), int64(1), object(7)
memory usage: 3.9+ MB


In [615]:
# Checking number of null values present in columns
df1.isnull().sum()

online_order                 0
book_table                   0
rate                         0
votes                        0
rest_type                    0
dish_liked                   0
cuisines                     0
approx_costfor_two_people    0
listed_intype                0
listed_incity                0
dtype: int64

In [616]:
# Describing the data
df1.describe().T

count        mean         std   min    25%  \
rate                       51717.0    3.700362    0.395391   1.8    3.5   
votes                      51717.0  283.697527  803.838853   0.0    7.0   
approx_costfor_two_people  51717.0  554.391689  437.563723  40.0  300.0   

                             50%    75%      max  
rate                         3.7    3.9      4.9  
votes                       41.0  198.0  16832.0  
approx_costfor_two_people  400.0  650.0   6000.0

Step 5: Binary Encoding

Step 5.1: online_order

In [617]:
# Checking unique values from column 'online_order'
print('Unique values from \'online_order\' column is', df1['online_order'].unique())

Unique values from 'online_order' column is ['Yes' 'No']


In [618]:
# Checking head/tail of the column
df1['online_order'].head()

0    Yes
1    Yes
2    Yes
3     No
4     No
Name: online_order, dtype: object

In [619]:
# Doing simple mapping for binary encoding
df1['online_order'] = df1['online_order'].map({'Yes': 1, 'No': 0})

In [620]:
# Checking head/tail of the column
df1['online_order'].head()

0    1
1    1
2    1
3    0
4    0
Name: online_order, dtype: int64

In [621]:
# Checking Info of the data
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   online_order               51717 non-null  int64  
 1   book_table                 51717 non-null  object 
 2   rate                       51717 non-null  float64
 3   votes                      51717 non-null  int64  
 4   rest_type                  51717 non-null  object 
 5   dish_liked                 51717 non-null  object 
 6   cuisines                   51717 non-null  object 
 7   approx_costfor_two_people  51717 non-null  float64
 8   listed_intype              51717 non-null  object 
 9   listed_incity              51717 non-null  object 
dtypes: float64(2), int64(2), object(6)
memory usage: 3.9+ MB


In [622]:
# Checking number of null values present in columns
df1.isnull().sum()

online_order                 0
book_table                   0
rate                         0
votes                        0
rest_type                    0
dish_liked                   0
cuisines                     0
approx_costfor_two_people    0
listed_intype                0
listed_incity                0
dtype: int64

In [623]:
# Describing the data
df1.describe().T

count        mean         std   min    25%  \
online_order               51717.0    0.588665    0.492080   0.0    0.0   
rate                       51717.0    3.700362    0.395391   1.8    3.5   
votes                      51717.0  283.697527  803.838853   0.0    7.0   
approx_costfor_two_people  51717.0  554.391689  437.563723  40.0  300.0   

                             50%    75%      max  
online_order                 1.0    1.0      1.0  
rate                         3.7    3.9      4.9  
votes                       41.0  198.0  16832.0  
approx_costfor_two_people  400.0  650.0   6000.0

Step 5.2: book_table

In [624]:
# Checking unique values from column 'book_table'
print('Unique values from \'book_table\' column is', df1['book_table'].unique())

Unique values from 'book_table' column is ['Yes' 'No']


In [625]:
# Checking head/tail of the column
df1['book_table'].head()

0    Yes
1     No
2     No
3     No
4     No
Name: book_table, dtype: object

In [626]:
# Doing simple mapping for binary encoding
df1['book_table'] = df1['book_table'].map({'Yes': 1, 'No': 0})

In [627]:
# Checking head/tail of the column
df1['book_table'].head()

0    1
1    0
2    0
3    0
4    0
Name: book_table, dtype: int64

In [628]:
# Checking Info of the data
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   online_order               51717 non-null  int64  
 1   book_table                 51717 non-null  int64  
 2   rate                       51717 non-null  float64
 3   votes                      51717 non-null  int64  
 4   rest_type                  51717 non-null  object 
 5   dish_liked                 51717 non-null  object 
 6   cuisines                   51717 non-null  object 
 7   approx_costfor_two_people  51717 non-null  float64
 8   listed_intype              51717 non-null  object 
 9   listed_incity              51717 non-null  object 
dtypes: float64(2), int64(3), object(5)
memory usage: 3.9+ MB


In [629]:
# Checking number of null values present in columns
df1.isnull().sum()

online_order                 0
book_table                   0
rate                         0
votes                        0
rest_type                    0
dish_liked                   0
cuisines                     0
approx_costfor_two_people    0
listed_intype                0
listed_incity                0
dtype: int64

In [630]:
# Describing the data
df1.describe().T

count        mean         std   min    25%  \
online_order               51717.0    0.588665    0.492080   0.0    0.0   
book_table                 51717.0    0.124698    0.330379   0.0    0.0   
rate                       51717.0    3.700362    0.395391   1.8    3.5   
votes                      51717.0  283.697527  803.838853   0.0    7.0   
approx_costfor_two_people  51717.0  554.391689  437.563723  40.0  300.0   

                             50%    75%      max  
online_order                 1.0    1.0      1.0  
book_table                   0.0    0.0      1.0  
rate                         3.7    3.9      4.9  
votes                       41.0  198.0  16832.0  
approx_costfor_two_people  400.0  650.0   6000.0

Step 6: Data Type Conversion

Step 6.1: rate → float

In [631]:
# Checking type of 'rate' column
print('Type of \'rate\' column is :', df1['rate'].dtype)

Type of 'rate' column is : float64


Inference : 'rate' column is already in float dtype but as per instruction, I am doing conversion.

In [632]:
# conversing 'rate' column dtype into float dtype
df1['rate'] = df1['rate'].astype(float)

In [633]:
# Checking type of 'rate' column
print('Type of \'rate\' column is :', df1['rate'].dtype)

Type of 'rate' column is : float64


In [634]:
# Checking Info of the data
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   online_order               51717 non-null  int64  
 1   book_table                 51717 non-null  int64  
 2   rate                       51717 non-null  float64
 3   votes                      51717 non-null  int64  
 4   rest_type                  51717 non-null  object 
 5   dish_liked                 51717 non-null  object 
 6   cuisines                   51717 non-null  object 
 7   approx_costfor_two_people  51717 non-null  float64
 8   listed_intype              51717 non-null  object 
 9   listed_incity              51717 non-null  object 
dtypes: float64(2), int64(3), object(5)
memory usage: 3.9+ MB


In [635]:
# Checking number of null values present in columns
df1.isnull().sum()

online_order                 0
book_table                   0
rate                         0
votes                        0
rest_type                    0
dish_liked                   0
cuisines                     0
approx_costfor_two_people    0
listed_intype                0
listed_incity                0
dtype: int64

In [636]:
# Describing the data
df1.describe().T

count        mean         std   min    25%  \
online_order               51717.0    0.588665    0.492080   0.0    0.0   
book_table                 51717.0    0.124698    0.330379   0.0    0.0   
rate                       51717.0    3.700362    0.395391   1.8    3.5   
votes                      51717.0  283.697527  803.838853   0.0    7.0   
approx_costfor_two_people  51717.0  554.391689  437.563723  40.0  300.0   

                             50%    75%      max  
online_order                 1.0    1.0      1.0  
book_table                   0.0    0.0      1.0  
rate                         3.7    3.9      4.9  
votes                       41.0  198.0  16832.0  
approx_costfor_two_people  400.0  650.0   6000.0

Step 6.2: votes → integer

In [637]:
# Checking type of 'votes' column
print('Type of \'votes\' column is :', df1['votes'].dtype)

Type of 'votes' column is : int64


Inference : 'votes' column is already in int dtype but as per instruction, I am doing conversion.

In [638]:
# conversing 'votes' column dtype into int dtype
df1['votes'] = df1['votes'].astype(int)

In [639]:
# Checking type of 'votes' column
print('Type of \'votes\' column is :', df1['votes'].dtype)

Type of 'votes' column is : int64


In [640]:
# Checking Info of the data
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   online_order               51717 non-null  int64  
 1   book_table                 51717 non-null  int64  
 2   rate                       51717 non-null  float64
 3   votes                      51717 non-null  int64  
 4   rest_type                  51717 non-null  object 
 5   dish_liked                 51717 non-null  object 
 6   cuisines                   51717 non-null  object 
 7   approx_costfor_two_people  51717 non-null  float64
 8   listed_intype              51717 non-null  object 
 9   listed_incity              51717 non-null  object 
dtypes: float64(2), int64(3), object(5)
memory usage: 3.9+ MB


In [641]:
# Checking number of null values present in columns
df1.isnull().sum()

online_order                 0
book_table                   0
rate                         0
votes                        0
rest_type                    0
dish_liked                   0
cuisines                     0
approx_costfor_two_people    0
listed_intype                0
listed_incity                0
dtype: int64

In [642]:
# Describing the data
df1.describe().T

count        mean         std   min    25%  \
online_order               51717.0    0.588665    0.492080   0.0    0.0   
book_table                 51717.0    0.124698    0.330379   0.0    0.0   
rate                       51717.0    3.700362    0.395391   1.8    3.5   
votes                      51717.0  283.697527  803.838853   0.0    7.0   
approx_costfor_two_people  51717.0  554.391689  437.563723  40.0  300.0   

                             50%    75%      max  
online_order                 1.0    1.0      1.0  
book_table                   0.0    0.0      1.0  
rate                         3.7    3.9      4.9  
votes                       41.0  198.0  16832.0  
approx_costfor_two_people  400.0  650.0   6000.0

Step 6.3: approx_costfor_two_people → integer

In [643]:
# Checking type of 'approx_costfor_two_people' column
print('Type of \'approx_costfor_two_people\' column is :', df1['approx_costfor_two_people'].dtype)

Type of 'approx_costfor_two_people' column is : float64


In [644]:
# conversing 'approx_costfor_two_people' column dtype into int dtype
df1['approx_costfor_two_people'] = df1['approx_costfor_two_people'].astype(int)

In [645]:
# Checking type of 'raapprox_costfor_two_peoplete' column
print('Type of \'approx_costfor_two_people\' column is :', df1['approx_costfor_two_people'].dtype)

Type of 'approx_costfor_two_people' column is : int64


In [646]:
# Checking Info of the data
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   online_order               51717 non-null  int64  
 1   book_table                 51717 non-null  int64  
 2   rate                       51717 non-null  float64
 3   votes                      51717 non-null  int64  
 4   rest_type                  51717 non-null  object 
 5   dish_liked                 51717 non-null  object 
 6   cuisines                   51717 non-null  object 
 7   approx_costfor_two_people  51717 non-null  int64  
 8   listed_intype              51717 non-null  object 
 9   listed_incity              51717 non-null  object 
dtypes: float64(1), int64(4), object(5)
memory usage: 3.9+ MB


In [647]:
# Checking number of null values present in columns
df1.isnull().sum()

online_order                 0
book_table                   0
rate                         0
votes                        0
rest_type                    0
dish_liked                   0
cuisines                     0
approx_costfor_two_people    0
listed_intype                0
listed_incity                0
dtype: int64

In [648]:
# Describing the data
df1.describe().T

count        mean         std   min    25%  \
online_order               51717.0    0.588665    0.492080   0.0    0.0   
book_table                 51717.0    0.124698    0.330379   0.0    0.0   
rate                       51717.0    3.700362    0.395391   1.8    3.5   
votes                      51717.0  283.697527  803.838853   0.0    7.0   
approx_costfor_two_people  51717.0  554.391689  437.563723  40.0  300.0   

                             50%    75%      max  
online_order                 1.0    1.0      1.0  
book_table                   0.0    0.0      1.0  
rate                         3.7    3.9      4.9  
votes                       41.0  198.0  16832.0  
approx_costfor_two_people  400.0  650.0   6000.0

Dataset Merging for Mapping

In [649]:
# Merging df1 and df2 for location-based analysis
df = pd.merge(df1, df2, on='listed_incity', how='left')

In [650]:
# Checking head/tail of the dataset
df.head()

online_order  book_table  rate  votes            rest_type  \
0             1           1   4.1    775        Casual Dining   
1             1           0   4.1    787        Casual Dining   
2             1           0   3.8    918  Cafe, Casual Dining   
3             0           0   3.7     88          Quick Bites   
4             0           0   3.8    166        Casual Dining   

                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines  approx_costfor_two_people listed_intype  \
0  North Indian, Mughlai, Chinese                        800        Buffet   
1     Chinese, North Indian, Thai                        800        Buffet   
2          Cafe, Mexican, Italian                        800        Buffet   
3      South Indian, North Indian                        300        Buffet   
4        North Indian, Rajasthani                        600        Buffet   

  listed_incity   Latitude  Longitude  
0  Banashankari  12.939333  77.553982  
1  Banashankari  12.939333  77.553982  
2  Banashankari  12.939333  77.553982  
3  Banashankari  12.939333  77.553982  
4  Banashankari  12.939333  77.553982

Q.1) What is the shape of the given dataset?

In [651]:
# Checking shape of the dataset
df1.shape

(51717, 10)

Q.2)  How many restaurants serve North Indian cuisine?

In [652]:
# Counting how many restaurants serve North Indian cuisine
print('Number of restaurants serve North Indian cuisine is', df1[df1['cuisines'].str.contains('North Indian', case=False)].shape[0])

Number of restaurants serve North Indian cuisine is 21085


Q.3) What cuisine is most commonly offered by restaurants in Bangalore?

In [653]:
# Importing Counter for counting purpose
from collections import Counter

In [654]:
# First split it by ',' and then use counter and then most_common
cuisine_counts = Counter()
df['cuisines'].dropna().apply(lambda x: cuisine_counts.update([c.strip() for c in x.split(',')]))
most_common_cuisine = cuisine_counts.most_common(1)[0][0]

print('Most common food offered by Bangalore restaurants is', most_common_cuisine)

Most common food offered by Bangalore restaurants is North Indian


Q.4) Which locality in Bangalore has the highest average cost for dining (for two people)?

In [655]:
# Finding locality where cost for dining is highest....
print('Locality where cost for dining is highest is', df.groupby("listed_incity")["approx_costfor_two_people"].mean().sort_values(ascending=False).head(5))

Locality where cost for dining is highest is listed_incity
Church Street     770.361248
Brigade Road      766.676088
MG Road           759.364992
Lavelle Road      751.151950
Residency Road    739.573457
Name: approx_costfor_two_people, dtype: float64


Q.5) Which restaurant type has the top rating with over 1000 votes?

In [656]:
# Finding restaurant with more than 1000 votes and top rating
print('Restaurant with more than 1000 votes and high rating is', df[df["votes"] > 1000].groupby("rest_type")["rate"].mean().sort_values(ascending=False).head(5))

Restaurant with more than 1000 votes and high rating is rest_type
Bakery                  4.800000
Pub, Cafe               4.657500
Cafe, Dessert Parlor    4.578261
Dessert Parlor          4.518367
Microbrewery            4.508333
Name: rate, dtype: float64


Q.6) How much does it cost at minimum to eat out in Bangalore?

In [657]:
# Calculating minimum payout for dining in Bangalore
print('Minimum cost to eat out in Bangalore is', df['approx_costfor_two_people'].min())

Minimum cost to eat out in Bangalore is 40


In [658]:
# Checking answer is correct or not
df['approx_costfor_two_people'].unique()

array([ 800,  300,  600,  700,  550,  500,  450,  650,  400,  900,  200,
        750,  150,  850,  100, 1200,  350,  250,  950, 1000, 1500, 1300,
        199,   80, 1100,  160, 1600,  230,  130,   50,  190, 1700, 1400,
        180, 1350, 2200, 2000, 1800, 1900,  330, 2500, 2100, 3000, 2800,
       3400,   40, 1250, 3500, 4000, 2400, 2600,  120, 1450,  469,   70,
       3200,   60,  560,  240,  360, 6000, 1050, 2300, 4100, 5000, 3700,
       1650, 2700, 4500,  140])

Inference : Here minimum cost for dining out is 40 followed by 50, 60, 70, 80, 100, 120, 130, 140, 150, 160, 180, 190, 199, 200 etc. So choose 200 since from available option 200 is minimum.

Q.7) What percentage of total online orders is received by restaurants in Banashankari?

In [659]:
# Calculating % total online orders received by restaurants in Banashankari
total_online_orders = df[df["listed_incity"] == "Banashankari"].shape[0]
banashankari_orders = df[(df["listed_incity"] == "Banashankari") & (df["online_order"] == 1)].shape[0]
percentage = (banashankari_orders / total_online_orders) * 100
print('Percentage of total online orders received by restaurants in Banashankari is', percentage)

Percentage of total online orders received by restaurants in Banashankari is 63.26767091541136


Inference : Nearly, 63% orders are online. So choosing approx. option as 58%.

Q.8) Which locality has the most restaurants with over 500 votes and a rating below 3.0?

In [660]:
# Locality with most of restaurants with more than 500 votes and rating below 3.0
print('Locality with most of restaurants with more than 500 votes and rating below 3.0 is', df[(df["votes"] > 500) & (df["rate"] < 3.0)].groupby("listed_incity").size().sort_values(ascending=False).head(5))

Locality with most of restaurants with more than 500 votes and rating below 3.0 is listed_incity
Brookefield         8
Whitefield          6
Indiranagar         4
Old Airport Road    4
Marathahalli        3
dtype: int64


Q.9) Which locality in Bangalore should Zomato target for expansion based on restaurant type diversity?

In [661]:
# Locality in Bangalore, zomato should try for expansion based on restaurant type diversity
print('Locality in Bangalore, zomato should try for expansion based on restaurant type diversity is', df.groupby("listed_incity")["rest_type"].nunique().sort_values(ascending=False).head(5))

Locality in Bangalore, zomato should try for expansion based on restaurant type diversity is listed_incity
BTM              62
Brigade Road     61
Church Street    61
Lavelle Road     59
MG Road          59
Name: rest_type, dtype: int64


Q.10) What's the average cost difference between buffet and delivery restaurants?

In [662]:
# Finding average cost difference between buffet and delivery restaurants
buffet_avg = df[df["listed_intype"] == "Buffet"]["approx_costfor_two_people"].mean()
delivery_avg = df[df["listed_intype"] == "Delivery"]["approx_costfor_two_people"].mean()
diff = abs(buffet_avg - delivery_avg)
print('Average cost difference between buffet and delivery restaurants is', diff)

Average cost difference between buffet and delivery restaurants is 831.2513659898209


Q.11) What is the maximum number of votes received by any restaurant with online ordering?

In [663]:
# Calculating maximum number of votes received by any restaurant in online orderes
print('Maximum number of votes received by any restaurant with online ordering is', df[df["online_order"] == 1]["votes"].max())

Maximum number of votes received by any restaurant with online ordering is 16832


Q.12) What is the average rating of restaurants that serve both North Indian and Chinese cuisines?

In [664]:
# Avearge rating of restaurants who serves both north indian and chinese cuisines
both_cuisines = df[df["cuisines"].str.contains("North Indian") & df["cuisines"].str.contains("Chinese")]
print('Avearge rating of restaurants who serves both north indian and chinese cuisines is', both_cuisines["rate"].mean())

Avearge rating of restaurants who serves both north indian and chinese cuisines is 3.5919080217658412


Q.13)  What is the most profitable area for Zomato based on potential revenue estimation?

In [665]:
# Finding most profitable area for Zomato based on potential revenue estimation
df["revenue_estimate"] = df["approx_costfor_two_people"] * df["votes"]
df.groupby("listed_incity")["revenue_estimate"].sum().sort_values(ascending=False).head(5)

listed_incity
Koramangala 7th Block    1006195610
Koramangala 5th Block     987609890
MG Road                   926969270
BTM                       924636830
Koramangala 4th Block     919922860
Name: revenue_estimate, dtype: int64

Q.14) If Zomato wants to reduce customer complaints, which restaurant type should they focus on?

In [666]:
# Solution to reduce customer complaints
print('Solution to reduce customer complaints is', df[df["rate"] < 3.0].groupby("rest_type").size().sort_values(ascending=False).head(5))

Solution to reduce customer complaints is rest_type
Quick Bites           738
Casual Dining         730
Cafe                  182
Delivery              115
Takeaway, Delivery     93
dtype: int64


Q.15) In which area should Zomato invest by considering high rating (rate > 4.2), high number of votes (> 500) and including online orders?

In [667]:
# Finding that area
print('In which area should Zomato invest by considering high rating (rate > 4.2), high number of votes (> 500) and including online orders?',
      df[(df["rate"] > 4.2) & (df["votes"] > 500) & (df["online_order"] == 1)].groupby("listed_incity").size().sort_values(ascending=False).head(5))

In which area should Zomato invest by considering high rating (rate > 4.2), high number of votes (> 500) and including online orders? listed_incity
Koramangala 7th Block    97
MG Road                  97
Church Street            95
Koramangala 4th Block    95
Brigade Road             94
dtype: int64


Cuisine-Specific Mapping using Folium Library.

In [668]:
from folium.plugins import MarkerCluster
import folium
from IPython.display import IFrame

In [669]:
italian_df = df[df['cuisines'].str.contains('Italian', na=False)]

In [670]:
# Approximate center of Bangalore
bangalore_center = [12.9716, 77.5946]

# Create base map
italian_map = folium.Map(location=bangalore_center, zoom_start=12)

In [ ]:
for _, row in italian_df.iterrows():
    if pd.notnull(row['Latitude']) and pd.notnull(row['Longitude']):
        popup_text = f"""
        <b>City:</b> {row['listed_incity']}<br>
        <b>Cuisines:</b> {row['cuisines']}
        """
        folium.Marker(
            location=[row['Latitude'], row['Longitude']],
            popup=popup_text,
            icon=folium.Icon(color='purple', icon='cutlery', prefix='fa')
        ).add_to(italian_map)

    italian_map.save("restaurant_density.html")

In [ ]:
IFrame("restaurant_density.html", width=700, height=500)

Interactive Density Mapping using Folium

In [ ]:
# Bangalore's approximate center coordinates
bangalore_center = [12.9716, 77.5946]

# Create the base map
restaurant_map = folium.Map(location=bangalore_center, zoom_start=12)

In [ ]:
marker_cluster = MarkerCluster().add_to(restaurant_map)

In [ ]:
for _, row in df.iterrows():
    if pd.notnull(row['Latitude']) and pd.notnull(row['Longitude']):
        popup_text = f"""
        <b>City:</b> {row['listed_incity']}<br>
        <b>Rating:</b> {row['rate']}<br>
        <b>Cost for Two:</b> ₹{row['approx_costfor_two_people']}
        """
        folium.Marker(
            location=[row['Latitude'], row['Longitude']],
            popup=popup_text,
            icon=folium.Icon(color='blue', icon='cutlery', prefix='fa')
        ).add_to(marker_cluster)

restaurant_map.save("restaurant_density.html")

ERROR:root:Internal Python error in the inspect module.
Below is the traceback from this internal error.

ERROR:root:Internal Python error in the inspect module.
Below is the traceback from this internal error.

ERROR:root:Internal Python error in the inspect module.
Below is the traceback from this internal error.



Traceback (most recent call last):
  File "/usr/local/lib/python3.11/dist-packages/IPython/core/interactiveshell.py", line 3553, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-545-d46018981afc>", line 14, in <cell line: 0>
    restaurant_map.save("restaurant_density.html")
  File "/usr/local/lib/python3.11/dist-packages/branca/element.py", line 228, in save
    html = root.render(**kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/branca/element.py", line 405, in render
    child.render(**kwargs)
  File "/usr/local/lib/python3.11/dist-packages/folium/folium.py", line 417, in render
    super().render(**kwargs)
  File "/usr/local/lib/python3.11/dist-packages/folium/elements.py", line 33, in render
    super().render(**kwargs)
  File "/usr/local/lib/python3.11/dist-packages/branca/element.py", line 736, in render
    element.render(**kwargs)
  File "/usr/local/lib/python3.11/dist-packages/folium/elements.py"

In [ ]:
IFrame("restaurant_density.html", width=700, height=500)